### ISSUE TITLE
no implicit conversion of nil into String when using inspec with kitchen --concurrency N

### ISSUE THREAD

**Issue by stefanandres** [inspec/inspec#1598]
### Description

When using inspec with kitchen test --concurrency N (N being >1) often (randomly) some verify actions fail. Retrying will let other tasks fail each time.

All my test suits are using flat profiles (so no profiles at all)
e.g. webhooks.rb:
```
describe command('/opt/gitlabwebhooks/start; sleep 1') do
  its('exit_status') { should eq 0 }
end

jsonbody='{"ref": "refs/heads/production", "user_id": 12}'

describe command("curl http://localhost:8080/ -H 'X-Gitlab-Token: mysecrettoken' -X POST -H 'Content-Type: application/json' -d '#{jsonbody}'") do
    its('exit_status') { should eq 0 }
end

describe file('/tmp/myhook.log') do
  it { should exist }
  it { should be_file }
  its('content') { should match(/foo=bar/) }
  its('content') { should match(/ref=refs\/heads\/production/) }
end
```

### InSpec and Platform Version

Ubuntu-16.04
InSpec-0.17.0

### Stacktrace

```
16:28:12 >>>>>> ------Exception-------
16:28:12 >>>>>> Class: Kitchen::ActionFailed
16:28:12 >>>>>> Message: 10 actions failed.
16:28:12 >>>>>>     Verify failed on instance <gitlabwebhooks-ubuntu-1404>.  Please see .kitchen/logs/gitlabwebhooks-ubuntu-1404.log for more details
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on sys11file-ubuntu-1404
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on nagios-ubuntu-1404
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on nagios-ubuntu-1604
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on role-admin-ubuntu-1604
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on sys11nodejs-ubuntu-1604
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on redis-ubuntu-1604
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on sys11logstash-ubuntu-1404
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on php56-ubuntu-1604
16:28:12 >>>>>>     Failed to complete #verify action: [no implicit conversion of nil into String] on php70-ubuntu-1404
16:28:12 >>>>>> ----------------------
16:28:12 >>>>>> Please see .kitchen/logs/kitchen.log for more details
16:28:12 >>>>>> Also try running `kitchen diagnose --all` for configuration
```

kitchen-log:
````

E, [2017-03-23T16:18:32.684600 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: ------Exception-------
E, [2017-03-23T16:18:32.685073 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: Class: Kitchen::ActionFailed
E, [2017-03-23T16:18:32.686917 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: Message: Inspec Runner returns 1
E, [2017-03-23T16:18:32.688477 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: ----------------------
E, [2017-03-23T16:18:32.692938 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: ------Backtrace-------
E, [2017-03-23T16:18:32.692972 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/kitchen-inspec-0.17.0/lib/kitchen/verifier/inspec.rb:96:in `call'
E, [2017-03-23T16:18:32.692991 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:434:in `block in verify_action'
E, [2017-03-23T16:18:32.693005 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:527:in `synchronize_or_call'
E, [2017-03-23T16:18:32.693018 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:489:in `block in action'
E, [2017-03-23T16:18:32.693046 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /usr/lib/ruby/2.3.0/benchmark.rb:293:in `measure'
E, [2017-03-23T16:18:32.693058 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:488:in `action'
E, [2017-03-23T16:18:32.693523 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:426:in `verify_action'
E, [2017-03-23T16:18:32.693555 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:359:in `block in transition_to'
E, [2017-03-23T16:18:32.693588 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:358:in `each'
E, [2017-03-23T16:18:32.693602 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:358:in `transition_to'
E, [2017-03-23T16:18:32.734103 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:157:in `verify'
E, [2017-03-23T16:18:32.838021 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:186:in `block in test'
E, [2017-03-23T16:18:32.840158 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /usr/lib/ruby/2.3.0/benchmark.rb:293:in `measure'
E, [2017-03-23T16:18:32.859658 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/instance.rb:182:in `test'
E, [2017-03-23T16:18:32.881721 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/command.rb:197:in `public_send'
E, [2017-03-23T16:18:32.882137 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/command.rb:197:in `run_action_in_thread'
E, [2017-03-23T16:18:32.882166 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: /var/lib/gems/2.3.0/gems/test-kitchen-1.16.0/lib/kitchen/command.rb:169:in `block (2 levels) in run_action'
E, [2017-03-23T16:18:32.882180 #30310] ERROR -- gitlabwebhooks-ubuntu-1404: ----End Backtrace-----
```

**Comment by stefanandres** [comment]
Is there any way I can try to workaround that?

**Comment by adamleff** [comment]
@stefanandres can you please run a single kitchen instance with `-l debug`, for example:

```
kitchen verify nagios-ubuntu-1604 -l debug
```

... and provide the output in a gist? I'm unable to reproduce this or see where the error is with the information provided.

Thanks!

**Comment by stefanandres** [comment]
Suer thing! I've added the debug log, as well as the inspec.rb for this specific test to https://gist.github.com/stefanandres/3cf3cd33577949c0aa1463df9da611dd

It runs fine with kitchen verify -c 1

Ah, this is even another error that I did not report yet, but this also only happens on concurrency.

**Comment by stefanandres** [comment]
Here is the gist for no implicit conversion:
https://gist.github.com/stefanandres/8679b5ec27c6f858a40e02bd41869911

**Comment by adamleff** [comment]
I can reproduce this, and it's super-weird. I'm going to try digging into it.

**Comment by adamleff** [comment]
OK, I think figured out what's going on here.

RSpec does some unholy things globally in order for it to do what it does, and InSpec uses RSpec under the covers. Even though each of the concurrent verify jobs runs in a different thread, RSpec does its global state munging thing... so each InSpec runner actually sees the "profile" (even if we make one up if you're using flat spec files and not an InSpec profile format) for all the suites.

When you use flat files and not a formal InSpec Profile format, InSpec creates a profile under the covers for you, and names it something like "tests from `/path/to/tests`". When the formatter (the thing that prints out the test results) goes to do its thing, it has to match up each of the tests with the profile it came from. The way it does that today is by grabbing the profile name out of the test and then finding the profile by name.

If there are two profiles loaded with the same name (which will be the case if there are multiple concurrent Kitchen runs happening using the same tests), InSpec could hand back the wrong profile. Then when it can't find the control/test in the profile, it returns `nil` and you get this ugly error.

To solve this, I'll put up a PR that locates the profile by profile name AND control ID which will eliminate this symptom. Ultimately, we need to solve parallelism which will likely involve not using RSpec, but that's a problem for a different day.

/cc @arlimus @chris-rock

**Comment by stefanandres** [comment]
Thanks a lot for the effort :)

**Comment by dragon788** [comment]
If we switch from serverspec to InSpec will that fix our issue here and and the chef/kitchen#135 trouble? We currently have a mix of both across all our repos.
